cli/flox-rust-sdk/src/providers/build.rs

```diff
@@ -1010,6 +1010,127 @@ mod tests {
         );
     }
 
+    #[test]
+    fn build_impure_patches_binaries() {
+        let package_name = String::from("foo");
+        let file_name = String::from("links-against-libc");
+
+        // Force CGO without providing `gcc` from the Flox environment.
+        //
+        // In reality `gcc` and `libc` will come from the `devShell` rather than
+        // a host OS which complicates these tests and means that we can never
+        // trust `rpath` or `runpath`.
+        let manifest = formatdoc! {r#"
+            version = 1
+
+            [install]
+            go.pkg-path = "go"
+
+            [vars]
+            CGO_ENABLED = "1"
+
+            [build.{package_name}]
+            command = """
+                go build main.go
+                mkdir -p $out/bin
+                cp main $out/bin/{file_name}
+            """
+        "#};
+
+        let (mut flox, _temp_dir_handle) = flox_instance();
+        let mut env = new_path_environment(&flox, &manifest);
+        let env_path = env.parent_path().unwrap();
+
+        let libc_code = indoc! {r#"
+            package main
+
+            /*
+            #include <stdio.h>
+
+            void hello() {
+            printf("Hello from libc!\n");
+            }
+            */
+            import "C"
+
+            func main() {
+                C.hello()
+            }
+        "#};
+        fs::write(env_path.join("main.go"), libc_code).unwrap();
+
+        if let Client::Mock(ref mut client) = flox.catalog_client {
+            client.clear_and_load_responses_from_file("resolve/go.json");
+        } else {
+            panic!("expected Mock client")
+        };
+
+        assert_build_status(&flox, &mut env, &package_name, true);
+        let result_wrapped = result_dir(&env_path, &package_name)
+            .read_link() // store path
+            .unwrap()
+            .join("bin")
+            .join(format!(".{}-wrapped", &file_name));
+
+        // Read the binary without executing a linker like `ldd` does.
+        let bin = &fs::read(result_wrapped).unwrap();
+        let bin_object = goblin::Object::parse(bin).unwrap();
+        match bin_object {
+            // Linux
+            goblin::Object::Elf(elf) => {
+                // Should always come from the Flox environment.
+                // TODO: Gives false-positives when run from devShell.
+                let interpreter = elf.interpreter.unwrap();
+                assert!(
+                    interpreter.starts_with("/nix/store/"),
+                    "should patch interpreter to nix store, got {:?}",
+                    interpreter
+                );
```

Comment:
Open to ideas about how to make this more reliable.

Reply:
Could we differentiate between different libcs by using the full store path?

Reply:
I'm not sure how this would help. How would libc come from a different store path than the devShell?

Reply:
Still not completely sure I have my head wrapped around this. But if I tweak the test to install gcc@wrapper-12.2.0, revert the `autoPatchelfPostFixup`, and make this `assert!` and `assert_eq!` just to show what store path it is, I see `/nix/store/yaz7pyf0ah88g2v505l38n0f3wg2vzdj-glibc-2.37-8/lib/ld-linux-x86-64.so.2`

If I then undo reverting the patch, I see `/nix/store/m71p7f0nymb19yn1dascklyya2i96jfw-glibc-2.39-52/lib/ld-linux-x86-64.so.2`

So I think in the first case it's coming from gcc@wrapper-12.2.0, and in the second we're unconditionally patching to whatever is in nixpkgs rev `dc14ed...` pinned in build-manifest.nix

Looks like the devShell would probably bring the same as `dc14ed...` until https://github.com/flox/flox/pull/2231 (sidenote, should I have bumped the pin in that PR?)

We'll need to stop pinning nixpkgs in build-manifest.nix and instead use the same nixpkgs as the toplevel group (https://github.com/flox/flox/issues/2084), so maybe we could change the test to:
- Install gcc@old-version in package group other than toplevel (which prior to this patch would cause interpreter to be set to `/nix/store/yaz7pyf0ah88g2v505l38n0f3wg2vzdj-glibc-2.37-8/lib/ld-linux-x86-64.so.2`)
- Assert that interpreter is `/nix/store/m71p7f0nymb19yn1dascklyya2i96jfw-glibc-2.39-52/lib/ld-linux-x86-64.so.2` (which is correct as of this patch)
- When we do https://github.com/flox/flox/issues/2084, we can assert that the interpreter is set to whatever the toplevel group provides

Reply:
That sounds like it's doing the opposite of what we wanted to achieve by no longer using gcc/glibc from the environment when present. In fact, if I include `gcc@wrapper-12.2.0` and `autoPatchelfPostFixup`, then I end up with a binary that just hangs. So I think we:

1. shouldn't add `autoPatchelfPostFixup`
2. should the portability risks of impure builds in the documentation
3. can explore an `LD_AUDIT` warn mode separately
4. could still add a test to verify that a binary built with Go and GCC works as a regression test against adding `autoPatchelfPostFixup` but don't bother inspecting any attributes of the unwrapped binary